{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "70a77f9c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import os, sys\n",
    "import copy\n",
    "import socket\n",
    "from tqdm import tqdm\n",
    "import torch\n",
    "import pickle\n",
    "from torch import optim\n",
    "from torch.utils.tensorboard import SummaryWriter\n",
    "\n",
    "#!pip install networkx matplotlib\n",
    "import networkx as nx\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import sys, os\n",
    "sys.path.insert(0, os.path.abspath(os.path.join(os.getcwd(), \"../\")))\n",
    "from libs.topology_manager import *\n",
    "from libs import fl, nn, data, log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "8c9749cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save Logs To File (info | debug | warning | error | critical) [optional]\n",
    "log.init(\"info\")\n",
    "#log.init(\"info\", \"federated.log\")\n",
    "#log.init(\"debug\", \"flkafka.log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "76f22545",
   "metadata": {},
   "outputs": [],
   "source": [
    "adj_mat = np.array([[0,1,1,1,1,1],\n",
    "             [1,0,0,0,0,0],\n",
    "             [1,0,0,0,0,0],\n",
    "             [1,0,0,0,0,0],\n",
    "             [1,0,0,0,0,0], \n",
    "             [1,0,0,0,0,0]])\n",
    "\n",
    "node_types = {'aggregator': [0,1], 'trainer': [2,3], 'broadcaster': [4,5]}\n",
    "\n",
    "di_graph = nx.DiGraph(adj_mat)\n",
    "#nx.draw(di_graph)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "ea0623fc",
   "metadata": {},
   "outputs": [
    {
     "ename": "RuntimeError",
     "evalue": "can't start new thread",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-114-7afc0b1d2edb>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     13\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtb\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mSummaryWriter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'../out/runs/flkafka'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcomment\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"Mnist Distributed Federated training\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 15\u001b[0;31m \u001b[0mfedargs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mFedArgs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-114-7afc0b1d2edb>\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m     11\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mseed\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtopic\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'pyflx'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 13\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtb\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mSummaryWriter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'../out/runs/flkafka'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcomment\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"Mnist Distributed Federated training\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     14\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m \u001b[0mfedargs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mFedArgs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/syft/lib/python3.9/site-packages/torch/utils/tensorboard/writer.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, log_dir, comment, purge_step, max_queue, flush_secs, filename_suffix)\u001b[0m\n\u001b[1;32m    218\u001b[0m         \u001b[0;31m# and recreated later as needed.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    219\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfile_writer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mall_writers\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 220\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_file_writer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    221\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    222\u001b[0m         \u001b[0;31m# Create default bins for histograms, see generate_testdata.py in tensorflow/tensorboard\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/syft/lib/python3.9/site-packages/torch/utils/tensorboard/writer.py\u001b[0m in \u001b[0;36m_get_file_writer\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    248\u001b[0m         \u001b[0;34m\"\"\"Returns the default FileWriter instance. Recreates it if closed.\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    249\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mall_writers\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfile_writer\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 250\u001b[0;31m             self.file_writer = FileWriter(self.log_dir, self.max_queue,\n\u001b[0m\u001b[1;32m    251\u001b[0m                                           self.flush_secs, self.filename_suffix)\n\u001b[1;32m    252\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mall_writers\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfile_writer\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_logdir\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfile_writer\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/syft/lib/python3.9/site-packages/torch/utils/tensorboard/writer.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, log_dir, max_queue, flush_secs, filename_suffix)\u001b[0m\n\u001b[1;32m     58\u001b[0m         \u001b[0;31m# actually the ones passing in a PosixPath\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     59\u001b[0m         \u001b[0mlog_dir\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlog_dir\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 60\u001b[0;31m         self.event_writer = EventFileWriter(\n\u001b[0m\u001b[1;32m     61\u001b[0m             log_dir, max_queue, flush_secs, filename_suffix)\n\u001b[1;32m     62\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/syft/lib/python3.9/site-packages/tensorboard/summary/writer/event_file_writer.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, logdir, max_queue_size, flush_secs, filename_suffix)\u001b[0m\n\u001b[1;32m     86\u001b[0m         )  # noqa E128\n\u001b[1;32m     87\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_general_file_writer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mio\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgfile\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mGFile\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_file_name\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"wb\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 88\u001b[0;31m         self._async_writer = _AsyncWriter(\n\u001b[0m\u001b[1;32m     89\u001b[0m             \u001b[0mRecordWriter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_general_file_writer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmax_queue_size\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mflush_secs\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     90\u001b[0m         )\n",
      "\u001b[0;32m~/anaconda3/envs/syft/lib/python3.9/site-packages/tensorboard/summary/writer/event_file_writer.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, record_writer, max_queue_size, flush_secs)\u001b[0m\n\u001b[1;32m    158\u001b[0m         )\n\u001b[1;32m    159\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_lock\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mthreading\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mLock\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 160\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_worker\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstart\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    161\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    162\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mwrite\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbytestring\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/syft/lib/python3.9/threading.py\u001b[0m in \u001b[0;36mstart\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    872\u001b[0m             \u001b[0m_limbo\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    873\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 874\u001b[0;31m             \u001b[0m_start_new_thread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_bootstrap\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    875\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    876\u001b[0m             \u001b[0;32mwith\u001b[0m \u001b[0m_active_limbo_lock\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mRuntimeError\u001b[0m: can't start new thread"
     ]
    }
   ],
   "source": [
    "class FedArgs():\n",
    "    def __init__(self):\n",
    "        self.num_clients = len(adj_mat)\n",
    "        self.epochs = 10\n",
    "        self.local_rounds = 1\n",
    "        self.client_batch_size = 32\n",
    "        self.test_batch_size = 128\n",
    "        self.learning_rate = 1e-4\n",
    "        self.weight_decay = 1e-5\n",
    "        self.cuda = False\n",
    "        self.seed = 1\n",
    "        self.topic = 'pyflx'\n",
    "        self.tb = SummaryWriter('../out/runs/flkafka', comment=\"Mnist Distributed Federated training\")\n",
    "\n",
    "fedargs = FedArgs()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "6a3022ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "use_cuda = fedargs.cuda and torch.cuda.is_available()\n",
    "torch.manual_seed(fedargs.seed)\n",
    "device = torch.device(\"cuda\" if use_cuda else \"cpu\")\n",
    "kwargs = {\"num_workers\": 1, \"pin_memory\": True} if use_cuda else {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "3e8de969",
   "metadata": {},
   "outputs": [],
   "source": [
    "host = socket.gethostname()\n",
    "clients = [host + \"(\" + str(client + 1) + \")\" for client in range(fedargs.num_clients)]\n",
    "ctp = DistributedTopology(adj_mat, clients, node_types)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "e8a71374",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Initialize Global and Client models\n",
    "global_model = nn.ModelMNIST()\n",
    "client_models = {client: copy.deepcopy(global_model) for client in clients}\n",
    "\n",
    "# Function for training\n",
    "def train_model(model, train_loader, fedargs, device):\n",
    "    model, loss = fl.client_update(model,\n",
    "                                train_loader,\n",
    "                                fedargs.learning_rate,\n",
    "                                fedargs.weight_decay,\n",
    "                                fedargs.local_rounds,\n",
    "                                device)\n",
    "    return model, loss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "b7283a3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load MNIST Data to clients\n",
    "train_data, test_data = data.load_dataset(\"mnist\")\n",
    "clients_data = data.split_data(train_data, clients)\n",
    "client_train_loaders, client_test_loaders = data.load_client_data(clients_data, fedargs.client_batch_size, 0.2, **kwargs)\n",
    "test_loader = torch.utils.data.DataLoader(test_data, batch_size=fedargs.test_batch_size, shuffle=True, **kwargs)\n",
    "\n",
    "clients_info = {\n",
    "        client: {\"train_loader\": client_train_loaders[client],\n",
    "                 \"test_loader\": client_test_loaders[client]}\n",
    "        for client in clients\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "280fc803",
   "metadata": {},
   "outputs": [],
   "source": [
    "import asyncio\n",
    "import time\n",
    "\n",
    "def background(f):\n",
    "    def wrapped(*args, **kwargs):\n",
    "        return asyncio.get_event_loop().run_in_executor(None, f, *args, **kwargs)\n",
    "\n",
    "    return wrapped\n",
    "\n",
    "@background\n",
    "def process(client, epoch, ctp, model, train_loader, test_loader, fedargs, device):\n",
    "    log.info(\"Processing Client {}\".format(client))\n",
    "    \n",
    "    # Consume and Average, epoch passed is actually prev epoch, for which we want to consume updates\n",
    "    rcvd_models = ctp.consume_model(client, fedargs.topic, model, epoch - 1)\n",
    "    log.info(\"Client {} received {} model updates from {}\".format(client, len(rcvd_models), list(rcvd_models.keys())))\n",
    "    if len(rcvd_models) != 0:\n",
    "        model = fl.federated_avg(rcvd_models)\n",
    "\n",
    "    # Train    \n",
    "    model, loss = train_model(model, train_loader, fedargs, device)\n",
    "    \n",
    "    # Publish\n",
    "    ctp.produce_model(client, fedargs.topic, model, epoch)\n",
    "\n",
    "    # Plot and Log\n",
    "    for local_epoch, loss in enumerate(list(loss.values())):\n",
    "        fedargs.tb.add_scalars(\"Training Loss/\" + client, {str(epoch): loss}, str(local_epoch + 1))\n",
    "\n",
    "    log.jsondebug(loss, \"Epoch {} of {} : Federated Training loss, Client {}\".format(epoch, fedargs.epochs, client))\n",
    "    log.modeldebug(model, \"Epoch {} of {} : Client {} Update\".format(epoch, fedargs.epochs, client))\n",
    "\n",
    "    # Test\n",
    "    test_output = fl.eval(model, test_loader, device)\n",
    "    fedargs.tb.add_scalar(\"Accuracy/\" + client, test_output[\"accuracy\"], epoch)\n",
    "    log.jsoninfo(test_output, \"Test Outut after Epoch {} of {} for Client {}\".format(epoch, fedargs.epochs, client))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "fc40c00d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/10 [00:00<?, ?it/s]2021-08-16 23:19:20,499 - <ipython-input-110-c8ea6acd90e4>::<module>(l:8) : Federated Training Epoch 1 of 10 [MainProcess : MainThread (INFO)]\n",
      "2021-08-16 23:19:20,538 - <ipython-input-109-7a5cd7c1f6e3>::process(l:12) : Processing Client bladecluster.iitp.org(3) [MainProcess : asyncio_7 (INFO)]\n",
      "2021-08-16 23:19:20,548 - <ipython-input-109-7a5cd7c1f6e3>::process(l:12) : Processing Client bladecluster.iitp.org(2) [MainProcess : asyncio_11 (INFO)]\n",
      "2021-08-16 23:19:20,553 - <ipython-input-109-7a5cd7c1f6e3>::process(l:12) : Processing Client bladecluster.iitp.org(4) [MainProcess : asyncio_8 (INFO)]\n",
      "2021-08-16 23:19:20,558 - <ipython-input-109-7a5cd7c1f6e3>::process(l:12) : Processing Client bladecluster.iitp.org(1) [MainProcess : asyncio_6 (INFO)]\n",
      "2021-08-16 23:19:20,564 - <ipython-input-109-7a5cd7c1f6e3>::process(l:12) : Processing Client bladecluster.iitp.org(6) [MainProcess : asyncio_9 (INFO)]\n",
      "2021-08-16 23:19:20,578 - <ipython-input-109-7a5cd7c1f6e3>::process(l:12) : Processing Client bladecluster.iitp.org(5) [MainProcess : asyncio_10 (INFO)]\n",
      "2021-08-16 23:19:23,613 - /home/harsh_1921cs01/hub/F3IA/fl/libs/topology_manager.py::consume_model(l:101) : Exception KafkaError{code=UNKNOWN_TOPIC_OR_PART,val=3,str=\"Subscribed topic not available: pyflx: Broker: Unknown topic or partition\"} occured consuming update for client bladecluster.iitp.org(2) [MainProcess : asyncio_11 (ERROR)]\n",
      "2021-08-16 23:19:23,636 - /home/harsh_1921cs01/hub/F3IA/fl/libs/topology_manager.py::consume_model(l:101) : Exception KafkaError{code=UNKNOWN_TOPIC_OR_PART,val=3,str=\"Subscribed topic not available: pyflx: Broker: Unknown topic or partition\"} occured consuming update for client bladecluster.iitp.org(6) [MainProcess : asyncio_9 (ERROR)]\n",
      "2021-08-16 23:19:23,636 - /home/harsh_1921cs01/hub/F3IA/fl/libs/topology_manager.py::consume_model(l:101) : Exception KafkaError{code=UNKNOWN_TOPIC_OR_PART,val=3,str=\"Subscribed topic not available: pyflx: Broker: Unknown topic or partition\"} occured consuming update for client bladecluster.iitp.org(1) [MainProcess : asyncio_6 (ERROR)]\n",
      "2021-08-16 23:19:23,637 - /home/harsh_1921cs01/hub/F3IA/fl/libs/topology_manager.py::consume_model(l:101) : Exception KafkaError{code=UNKNOWN_TOPIC_OR_PART,val=3,str=\"Subscribed topic not available: pyflx: Broker: Unknown topic or partition\"} occured consuming update for client bladecluster.iitp.org(4) [MainProcess : asyncio_8 (ERROR)]\n",
      "2021-08-16 23:19:23,647 - /home/harsh_1921cs01/hub/F3IA/fl/libs/topology_manager.py::consume_model(l:101) : Exception KafkaError{code=UNKNOWN_TOPIC_OR_PART,val=3,str=\"Subscribed topic not available: pyflx: Broker: Unknown topic or partition\"} occured consuming update for client bladecluster.iitp.org(5) [MainProcess : asyncio_10 (ERROR)]\n",
      "2021-08-16 23:19:23,647 - /home/harsh_1921cs01/hub/F3IA/fl/libs/topology_manager.py::consume_model(l:101) : Exception KafkaError{code=UNKNOWN_TOPIC_OR_PART,val=3,str=\"Subscribed topic not available: pyflx: Broker: Unknown topic or partition\"} occured consuming update for client bladecluster.iitp.org(3) [MainProcess : asyncio_7 (ERROR)]\n",
      "2021-08-16 23:19:23,663 - <ipython-input-109-7a5cd7c1f6e3>::process(l:16) : Client bladecluster.iitp.org(6) received 0 model updates from [] [MainProcess : asyncio_9 (INFO)]\n",
      "2021-08-16 23:19:23,665 - <ipython-input-109-7a5cd7c1f6e3>::process(l:16) : Client bladecluster.iitp.org(2) received 0 model updates from [] [MainProcess : asyncio_11 (INFO)]\n",
      "2021-08-16 23:19:23,688 - <ipython-input-109-7a5cd7c1f6e3>::process(l:16) : Client bladecluster.iitp.org(4) received 0 model updates from [] [MainProcess : asyncio_8 (INFO)]\n",
      "2021-08-16 23:19:23,689 - <ipython-input-109-7a5cd7c1f6e3>::process(l:16) : Client bladecluster.iitp.org(1) received 0 model updates from [] [MainProcess : asyncio_6 (INFO)]\n",
      "2021-08-16 23:19:23,702 - <ipython-input-109-7a5cd7c1f6e3>::process(l:16) : Client bladecluster.iitp.org(5) received 0 model updates from [] [MainProcess : asyncio_10 (INFO)]\n",
      "2021-08-16 23:19:23,712 - <ipython-input-109-7a5cd7c1f6e3>::process(l:16) : Client bladecluster.iitp.org(3) received 0 model updates from [] [MainProcess : asyncio_7 (INFO)]\n",
      "2021-08-16 23:19:40,476 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:56) : Producing user records to topic pyflx. ^C to exit. [MainProcess : asyncio_11 (INFO)]\n",
      "2021-08-16 23:19:40,869 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:66) : Flushing records... [MainProcess : asyncio_11 (INFO)]\n",
      "2021-08-16 23:19:41,170 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:56) : Producing user records to topic pyflx. ^C to exit. [MainProcess : asyncio_9 (INFO)]\n",
      "2021-08-16 23:19:41,460 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:56) : Producing user records to topic pyflx. ^C to exit. [MainProcess : asyncio_8 (INFO)]\n",
      "2021-08-16 23:19:41,808 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:56) : Producing user records to topic pyflx. ^C to exit. [MainProcess : asyncio_6 (INFO)]\n",
      "2021-08-16 23:19:42,129 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:66) : Flushing records... [MainProcess : asyncio_8 (INFO)]\n",
      "2021-08-16 23:19:42,144 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:66) : Flushing records... [MainProcess : asyncio_9 (INFO)]\n",
      "2021-08-16 23:19:42,163 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:66) : Flushing records... [MainProcess : asyncio_6 (INFO)]\n",
      "2021-08-16 23:19:42,221 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::delivery_report(l:48) : Delivery failed for User record b'bladecluster.iitp.org(6)': KafkaError{code=MSG_SIZE_TOO_LARGE,val=10,str=\"Broker: Message size too large\"} [MainProcess : asyncio_9 (ERROR)]\n",
      "2021-08-16 23:19:42,536 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:56) : Producing user records to topic pyflx. ^C to exit. [MainProcess : asyncio_10 (INFO)]\n",
      "2021-08-16 23:19:42,556 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::delivery_report(l:48) : Delivery failed for User record b'bladecluster.iitp.org(4)': KafkaError{code=MSG_SIZE_TOO_LARGE,val=10,str=\"Broker: Message size too large\"} [MainProcess : asyncio_8 (ERROR)]\n",
      "2021-08-16 23:19:42,587 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::delivery_report(l:48) : Delivery failed for User record b'bladecluster.iitp.org(2)': KafkaError{code=MSG_SIZE_TOO_LARGE,val=10,str=\"Broker: Message size too large\"} [MainProcess : asyncio_11 (ERROR)]\n",
      "2021-08-16 23:19:42,799 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:66) : Flushing records... [MainProcess : asyncio_10 (INFO)]\n",
      "2021-08-16 23:19:43,148 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:56) : Producing user records to topic pyflx. ^C to exit. [MainProcess : asyncio_7 (INFO)]\n",
      "2021-08-16 23:19:43,163 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::delivery_report(l:48) : Delivery failed for User record b'bladecluster.iitp.org(1)': KafkaError{code=MSG_SIZE_TOO_LARGE,val=10,str=\"Broker: Message size too large\"} [MainProcess : asyncio_6 (ERROR)]\n",
      "2021-08-16 23:19:43,390 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:66) : Flushing records... [MainProcess : asyncio_7 (INFO)]\n",
      "2021-08-16 23:19:43,635 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::delivery_report(l:48) : Delivery failed for User record b'bladecluster.iitp.org(5)': KafkaError{code=MSG_SIZE_TOO_LARGE,val=10,str=\"Broker: Message size too large\"} [MainProcess : asyncio_10 (ERROR)]\n",
      "2021-08-16 23:19:44,232 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::delivery_report(l:48) : Delivery failed for User record b'bladecluster.iitp.org(3)': KafkaError{code=MSG_SIZE_TOO_LARGE,val=10,str=\"Broker: Message size too large\"} [MainProcess : asyncio_7 (ERROR)]\n",
      "2021-08-16 23:19:45,931 - <ipython-input-109-7a5cd7c1f6e3>::process(l:36) : Test Outut after Epoch 1 of 10 for Client bladecluster.iitp.org(4) {\n",
      "    \"accuracy\": 89.1,\n",
      "    \"attack\": {\n",
      "        \"attack_success_count\": 0,\n",
      "        \"attack_success_rate\": 0,\n",
      "        \"instances\": 0,\n",
      "        \"misclassification_rate\": 0,\n",
      "        \"misclassifications\": 0\n",
      "    },\n",
      "    \"correct\": 1782,\n",
      "    \"test_loss\": 0.3582343509197235\n",
      "} [MainProcess : asyncio_8 (INFO)]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-08-16 23:19:46,029 - <ipython-input-109-7a5cd7c1f6e3>::process(l:36) : Test Outut after Epoch 1 of 10 for Client bladecluster.iitp.org(2) {\n",
      "    \"accuracy\": 90.3,\n",
      "    \"attack\": {\n",
      "        \"attack_success_count\": 0,\n",
      "        \"attack_success_rate\": 0,\n",
      "        \"instances\": 0,\n",
      "        \"misclassification_rate\": 0,\n",
      "        \"misclassifications\": 0\n",
      "    },\n",
      "    \"correct\": 1806,\n",
      "    \"test_loss\": 0.3241649172306061\n",
      "} [MainProcess : asyncio_11 (INFO)]\n",
      "2021-08-16 23:19:46,045 - <ipython-input-109-7a5cd7c1f6e3>::process(l:36) : Test Outut after Epoch 1 of 10 for Client bladecluster.iitp.org(6) {\n",
      "    \"accuracy\": 90.35,\n",
      "    \"attack\": {\n",
      "        \"attack_success_count\": 0,\n",
      "        \"attack_success_rate\": 0,\n",
      "        \"instances\": 0,\n",
      "        \"misclassification_rate\": 0,\n",
      "        \"misclassifications\": 0\n",
      "    },\n",
      "    \"correct\": 1807,\n",
      "    \"test_loss\": 0.3523443686962128\n",
      "} [MainProcess : asyncio_9 (INFO)]\n",
      "2021-08-16 23:19:46,096 - <ipython-input-109-7a5cd7c1f6e3>::process(l:36) : Test Outut after Epoch 1 of 10 for Client bladecluster.iitp.org(1) {\n",
      "    \"accuracy\": 91.4,\n",
      "    \"attack\": {\n",
      "        \"attack_success_count\": 0,\n",
      "        \"attack_success_rate\": 0,\n",
      "        \"instances\": 0,\n",
      "        \"misclassification_rate\": 0,\n",
      "        \"misclassifications\": 0\n",
      "    },\n",
      "    \"correct\": 1828,\n",
      "    \"test_loss\": 0.35136747038364413\n",
      "} [MainProcess : asyncio_6 (INFO)]\n",
      "2021-08-16 23:19:46,265 - <ipython-input-109-7a5cd7c1f6e3>::process(l:36) : Test Outut after Epoch 1 of 10 for Client bladecluster.iitp.org(5) {\n",
      "    \"accuracy\": 89.64999999999999,\n",
      "    \"attack\": {\n",
      "        \"attack_success_count\": 0,\n",
      "        \"attack_success_rate\": 0,\n",
      "        \"instances\": 0,\n",
      "        \"misclassification_rate\": 0,\n",
      "        \"misclassifications\": 0\n",
      "    },\n",
      "    \"correct\": 1793,\n",
      "    \"test_loss\": 0.3470179681777954\n",
      "} [MainProcess : asyncio_10 (INFO)]\n",
      "2021-08-16 23:19:46,499 - <ipython-input-109-7a5cd7c1f6e3>::process(l:36) : Test Outut after Epoch 1 of 10 for Client bladecluster.iitp.org(3) {\n",
      "    \"accuracy\": 88.85,\n",
      "    \"attack\": {\n",
      "        \"attack_success_count\": 0,\n",
      "        \"attack_success_rate\": 0,\n",
      "        \"instances\": 0,\n",
      "        \"misclassification_rate\": 0,\n",
      "        \"misclassifications\": 0\n",
      "    },\n",
      "    \"correct\": 1777,\n",
      "    \"test_loss\": 0.3667033026218414\n",
      "} [MainProcess : asyncio_7 (INFO)]\n",
      " 10%|█         | 1/10 [00:26<03:54, 26.01s/it]2021-08-16 23:19:46,516 - <ipython-input-110-c8ea6acd90e4>::<module>(l:8) : Federated Training Epoch 2 of 10 [MainProcess : MainThread (INFO)]\n",
      "2021-08-16 23:19:46,539 - <ipython-input-109-7a5cd7c1f6e3>::process(l:12) : Processing Client bladecluster.iitp.org(1) [MainProcess : asyncio_3 (INFO)]\n",
      "2021-08-16 23:19:46,560 - <ipython-input-109-7a5cd7c1f6e3>::process(l:12) : Processing Client bladecluster.iitp.org(3) [MainProcess : asyncio_2 (INFO)]\n",
      "2021-08-16 23:19:46,579 - <ipython-input-109-7a5cd7c1f6e3>::process(l:12) : Processing Client bladecluster.iitp.org(6) [MainProcess : asyncio_1 (INFO)]\n",
      "2021-08-16 23:19:46,579 - <ipython-input-109-7a5cd7c1f6e3>::process(l:12) : Processing Client bladecluster.iitp.org(2) [MainProcess : asyncio_0 (INFO)]\n",
      "2021-08-16 23:19:46,580 - <ipython-input-109-7a5cd7c1f6e3>::process(l:12) : Processing Client bladecluster.iitp.org(5) [MainProcess : asyncio_5 (INFO)]\n",
      "2021-08-16 23:19:46,585 - <ipython-input-109-7a5cd7c1f6e3>::process(l:12) : Processing Client bladecluster.iitp.org(4) [MainProcess : asyncio_4 (INFO)]\n",
      "2021-08-16 23:19:56,586 - <ipython-input-109-7a5cd7c1f6e3>::process(l:16) : Client bladecluster.iitp.org(1) received 0 model updates from [] [MainProcess : asyncio_3 (INFO)]\n",
      "2021-08-16 23:19:56,615 - <ipython-input-109-7a5cd7c1f6e3>::process(l:16) : Client bladecluster.iitp.org(3) received 0 model updates from [] [MainProcess : asyncio_2 (INFO)]\n",
      "2021-08-16 23:19:56,649 - <ipython-input-109-7a5cd7c1f6e3>::process(l:16) : Client bladecluster.iitp.org(2) received 0 model updates from [] [MainProcess : asyncio_0 (INFO)]\n",
      "2021-08-16 23:19:56,665 - <ipython-input-109-7a5cd7c1f6e3>::process(l:16) : Client bladecluster.iitp.org(6) received 0 model updates from [] [MainProcess : asyncio_1 (INFO)]\n",
      "2021-08-16 23:19:56,670 - <ipython-input-109-7a5cd7c1f6e3>::process(l:16) : Client bladecluster.iitp.org(4) received 0 model updates from [] [MainProcess : asyncio_4 (INFO)]\n",
      "2021-08-16 23:19:56,671 - <ipython-input-109-7a5cd7c1f6e3>::process(l:16) : Client bladecluster.iitp.org(5) received 0 model updates from [] [MainProcess : asyncio_5 (INFO)]\n",
      "2021-08-16 23:20:15,828 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:56) : Producing user records to topic pyflx. ^C to exit. [MainProcess : asyncio_4 (INFO)]\n",
      "2021-08-16 23:20:16,063 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:66) : Flushing records... [MainProcess : asyncio_4 (INFO)]\n",
      "2021-08-16 23:20:16,529 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:56) : Producing user records to topic pyflx. ^C to exit. [MainProcess : asyncio_2 (INFO)]\n",
      "2021-08-16 23:20:16,808 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:56) : Producing user records to topic pyflx. ^C to exit. [MainProcess : asyncio_3 (INFO)]\n",
      "2021-08-16 23:20:16,891 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::delivery_report(l:48) : Delivery failed for User record b'bladecluster.iitp.org(4)': KafkaError{code=MSG_SIZE_TOO_LARGE,val=10,str=\"Broker: Message size too large\"} [MainProcess : asyncio_4 (ERROR)]\n",
      "2021-08-16 23:20:17,085 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:66) : Flushing records... [MainProcess : asyncio_2 (INFO)]\n",
      "2021-08-16 23:20:17,127 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:66) : Flushing records... [MainProcess : asyncio_3 (INFO)]\n",
      "2021-08-16 23:20:17,589 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::delivery_report(l:48) : Delivery failed for User record b'bladecluster.iitp.org(3)': KafkaError{code=MSG_SIZE_TOO_LARGE,val=10,str=\"Broker: Message size too large\"} [MainProcess : asyncio_2 (ERROR)]\n",
      "2021-08-16 23:20:17,589 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:56) : Producing user records to topic pyflx. ^C to exit. [MainProcess : asyncio_1 (INFO)]\n",
      "2021-08-16 23:20:17,779 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:66) : Flushing records... [MainProcess : asyncio_1 (INFO)]\n",
      "2021-08-16 23:20:17,859 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::delivery_report(l:48) : Delivery failed for User record b'bladecluster.iitp.org(1)': KafkaError{code=MSG_SIZE_TOO_LARGE,val=10,str=\"Broker: Message size too large\"} [MainProcess : asyncio_3 (ERROR)]\n",
      "2021-08-16 23:20:18,507 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:56) : Producing user records to topic pyflx. ^C to exit. [MainProcess : asyncio_5 (INFO)]\n",
      "2021-08-16 23:20:18,647 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::delivery_report(l:48) : Delivery failed for User record b'bladecluster.iitp.org(6)': KafkaError{code=MSG_SIZE_TOO_LARGE,val=10,str=\"Broker: Message size too large\"} [MainProcess : asyncio_1 (ERROR)]\n",
      "2021-08-16 23:20:18,721 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:66) : Flushing records... [MainProcess : asyncio_5 (INFO)]\n",
      "2021-08-16 23:20:19,152 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:56) : Producing user records to topic pyflx. ^C to exit. [MainProcess : asyncio_0 (INFO)]\n",
      "2021-08-16 23:20:19,305 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:66) : Flushing records... [MainProcess : asyncio_0 (INFO)]\n",
      "2021-08-16 23:20:19,608 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::delivery_report(l:48) : Delivery failed for User record b'bladecluster.iitp.org(5)': KafkaError{code=MSG_SIZE_TOO_LARGE,val=10,str=\"Broker: Message size too large\"} [MainProcess : asyncio_5 (ERROR)]\n",
      "2021-08-16 23:20:20,276 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::delivery_report(l:48) : Delivery failed for User record b'bladecluster.iitp.org(2)': KafkaError{code=MSG_SIZE_TOO_LARGE,val=10,str=\"Broker: Message size too large\"} [MainProcess : asyncio_0 (ERROR)]\n",
      "2021-08-16 23:20:20,434 - <ipython-input-109-7a5cd7c1f6e3>::process(l:36) : Test Outut after Epoch 2 of 10 for Client bladecluster.iitp.org(4) {\n",
      "    \"accuracy\": 89.7,\n",
      "    \"attack\": {\n",
      "        \"attack_success_count\": 0,\n",
      "        \"attack_success_rate\": 0,\n",
      "        \"instances\": 0,\n",
      "        \"misclassification_rate\": 0,\n",
      "        \"misclassifications\": 0\n",
      "    },\n",
      "    \"correct\": 1794,\n",
      "    \"test_loss\": 0.35239679980278016\n",
      "} [MainProcess : asyncio_4 (INFO)]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-08-16 23:20:20,728 - <ipython-input-109-7a5cd7c1f6e3>::process(l:36) : Test Outut after Epoch 2 of 10 for Client bladecluster.iitp.org(3) {\n",
      "    \"accuracy\": 91.05,\n",
      "    \"attack\": {\n",
      "        \"attack_success_count\": 0,\n",
      "        \"attack_success_rate\": 0,\n",
      "        \"instances\": 0,\n",
      "        \"misclassification_rate\": 0,\n",
      "        \"misclassifications\": 0\n",
      "    },\n",
      "    \"correct\": 1821,\n",
      "    \"test_loss\": 0.32294040703773497\n",
      "} [MainProcess : asyncio_2 (INFO)]\n",
      "2021-08-16 23:20:20,943 - <ipython-input-109-7a5cd7c1f6e3>::process(l:36) : Test Outut after Epoch 2 of 10 for Client bladecluster.iitp.org(1) {\n",
      "    \"accuracy\": 91.55,\n",
      "    \"attack\": {\n",
      "        \"attack_success_count\": 0,\n",
      "        \"attack_success_rate\": 0,\n",
      "        \"instances\": 0,\n",
      "        \"misclassification_rate\": 0,\n",
      "        \"misclassifications\": 0\n",
      "    },\n",
      "    \"correct\": 1831,\n",
      "    \"test_loss\": 0.3304774861335754\n",
      "} [MainProcess : asyncio_3 (INFO)]\n",
      "2021-08-16 23:20:21,175 - <ipython-input-109-7a5cd7c1f6e3>::process(l:36) : Test Outut after Epoch 2 of 10 for Client bladecluster.iitp.org(6) {\n",
      "    \"accuracy\": 90.75,\n",
      "    \"attack\": {\n",
      "        \"attack_success_count\": 0,\n",
      "        \"attack_success_rate\": 0,\n",
      "        \"instances\": 0,\n",
      "        \"misclassification_rate\": 0,\n",
      "        \"misclassifications\": 0\n",
      "    },\n",
      "    \"correct\": 1815,\n",
      "    \"test_loss\": 0.33108617401123047\n",
      "} [MainProcess : asyncio_1 (INFO)]\n",
      "2021-08-16 23:20:21,538 - <ipython-input-109-7a5cd7c1f6e3>::process(l:36) : Test Outut after Epoch 2 of 10 for Client bladecluster.iitp.org(5) {\n",
      "    \"accuracy\": 90.55,\n",
      "    \"attack\": {\n",
      "        \"attack_success_count\": 0,\n",
      "        \"attack_success_rate\": 0,\n",
      "        \"instances\": 0,\n",
      "        \"misclassification_rate\": 0,\n",
      "        \"misclassifications\": 0\n",
      "    },\n",
      "    \"correct\": 1811,\n",
      "    \"test_loss\": 0.34308029556274416\n",
      "} [MainProcess : asyncio_5 (INFO)]\n",
      "2021-08-16 23:20:21,795 - <ipython-input-109-7a5cd7c1f6e3>::process(l:36) : Test Outut after Epoch 2 of 10 for Client bladecluster.iitp.org(2) {\n",
      "    \"accuracy\": 90.2,\n",
      "    \"attack\": {\n",
      "        \"attack_success_count\": 0,\n",
      "        \"attack_success_rate\": 0,\n",
      "        \"instances\": 0,\n",
      "        \"misclassification_rate\": 0,\n",
      "        \"misclassifications\": 0\n",
      "    },\n",
      "    \"correct\": 1804,\n",
      "    \"test_loss\": 0.33905955743789673\n",
      "} [MainProcess : asyncio_0 (INFO)]\n",
      " 20%|██        | 2/10 [01:01<04:11, 31.47s/it]2021-08-16 23:20:21,813 - <ipython-input-110-c8ea6acd90e4>::<module>(l:8) : Federated Training Epoch 3 of 10 [MainProcess : MainThread (INFO)]\n",
      "2021-08-16 23:20:21,872 - <ipython-input-109-7a5cd7c1f6e3>::process(l:12) : Processing Client bladecluster.iitp.org(3) [MainProcess : asyncio_6 (INFO)]\n",
      "2021-08-16 23:20:21,880 - <ipython-input-109-7a5cd7c1f6e3>::process(l:12) : Processing Client bladecluster.iitp.org(5) [MainProcess : asyncio_7 (INFO)]\n",
      "2021-08-16 23:20:21,881 - <ipython-input-109-7a5cd7c1f6e3>::process(l:12) : Processing Client bladecluster.iitp.org(1) [MainProcess : asyncio_11 (INFO)]\n",
      "2021-08-16 23:20:21,888 - <ipython-input-109-7a5cd7c1f6e3>::process(l:12) : Processing Client bladecluster.iitp.org(2) [MainProcess : asyncio_9 (INFO)]\n",
      "2021-08-16 23:20:21,890 - <ipython-input-109-7a5cd7c1f6e3>::process(l:12) : Processing Client bladecluster.iitp.org(6) [MainProcess : asyncio_8 (INFO)]\n",
      "2021-08-16 23:20:21,890 - <ipython-input-109-7a5cd7c1f6e3>::process(l:12) : Processing Client bladecluster.iitp.org(4) [MainProcess : asyncio_10 (INFO)]\n",
      "2021-08-16 23:20:31,905 - <ipython-input-109-7a5cd7c1f6e3>::process(l:16) : Client bladecluster.iitp.org(3) received 0 model updates from [] [MainProcess : asyncio_6 (INFO)]\n",
      "2021-08-16 23:20:31,921 - <ipython-input-109-7a5cd7c1f6e3>::process(l:16) : Client bladecluster.iitp.org(5) received 0 model updates from [] [MainProcess : asyncio_7 (INFO)]\n",
      "2021-08-16 23:20:31,942 - <ipython-input-109-7a5cd7c1f6e3>::process(l:16) : Client bladecluster.iitp.org(6) received 0 model updates from [] [MainProcess : asyncio_8 (INFO)]\n",
      "2021-08-16 23:20:31,970 - <ipython-input-109-7a5cd7c1f6e3>::process(l:16) : Client bladecluster.iitp.org(1) received 0 model updates from [] [MainProcess : asyncio_11 (INFO)]\n",
      "2021-08-16 23:20:31,993 - <ipython-input-109-7a5cd7c1f6e3>::process(l:16) : Client bladecluster.iitp.org(4) received 0 model updates from [] [MainProcess : asyncio_10 (INFO)]\n",
      "2021-08-16 23:20:31,999 - <ipython-input-109-7a5cd7c1f6e3>::process(l:16) : Client bladecluster.iitp.org(2) received 0 model updates from [] [MainProcess : asyncio_9 (INFO)]\n",
      "2021-08-16 23:20:48,345 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:56) : Producing user records to topic pyflx. ^C to exit. [MainProcess : asyncio_11 (INFO)]\n",
      "2021-08-16 23:20:48,530 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:66) : Flushing records... [MainProcess : asyncio_11 (INFO)]\n",
      "2021-08-16 23:20:49,429 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::delivery_report(l:50) : User record b'bladecluster.iitp.org(1)' successfully produced to pyflx [0] at offset 0 [MainProcess : asyncio_11 (INFO)]\n",
      "2021-08-16 23:20:50,113 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:56) : Producing user records to topic pyflx. ^C to exit. [MainProcess : asyncio_10 (INFO)]\n",
      "2021-08-16 23:20:50,280 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:66) : Flushing records... [MainProcess : asyncio_10 (INFO)]\n",
      "2021-08-16 23:20:50,600 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:56) : Producing user records to topic pyflx. ^C to exit. [MainProcess : asyncio_9 (INFO)]\n",
      "2021-08-16 23:20:50,769 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:66) : Flushing records... [MainProcess : asyncio_9 (INFO)]\n",
      "2021-08-16 23:20:51,134 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:56) : Producing user records to topic pyflx. ^C to exit. [MainProcess : asyncio_6 (INFO)]\n",
      "2021-08-16 23:20:51,206 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::delivery_report(l:50) : User record b'bladecluster.iitp.org(4)' successfully produced to pyflx [0] at offset 1 [MainProcess : asyncio_10 (INFO)]\n",
      "2021-08-16 23:20:51,594 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:56) : Producing user records to topic pyflx. ^C to exit. [MainProcess : asyncio_8 (INFO)]\n",
      "2021-08-16 23:20:51,667 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:66) : Flushing records... [MainProcess : asyncio_6 (INFO)]\n",
      "2021-08-16 23:20:51,692 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::delivery_report(l:50) : User record b'bladecluster.iitp.org(2)' successfully produced to pyflx [0] at offset 2 [MainProcess : asyncio_9 (INFO)]\n",
      "2021-08-16 23:20:51,819 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:66) : Flushing records... [MainProcess : asyncio_8 (INFO)]\n",
      "2021-08-16 23:20:52,294 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:56) : Producing user records to topic pyflx. ^C to exit. [MainProcess : asyncio_7 (INFO)]\n",
      "2021-08-16 23:20:52,310 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::delivery_report(l:50) : User record b'bladecluster.iitp.org(3)' successfully produced to pyflx [0] at offset 3 [MainProcess : asyncio_6 (INFO)]\n",
      "2021-08-16 23:20:52,519 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:66) : Flushing records... [MainProcess : asyncio_7 (INFO)]\n",
      "2021-08-16 23:20:52,651 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::delivery_report(l:50) : User record b'bladecluster.iitp.org(6)' successfully produced to pyflx [0] at offset 4 [MainProcess : asyncio_8 (INFO)]\n",
      "2021-08-16 23:20:53,592 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::delivery_report(l:50) : User record b'bladecluster.iitp.org(5)' successfully produced to pyflx [0] at offset 5 [MainProcess : asyncio_7 (INFO)]\n",
      "2021-08-16 23:20:54,264 - <ipython-input-109-7a5cd7c1f6e3>::process(l:36) : Test Outut after Epoch 3 of 10 for Client bladecluster.iitp.org(1) {\n",
      "    \"accuracy\": 91.9,\n",
      "    \"attack\": {\n",
      "        \"attack_success_count\": 0,\n",
      "        \"attack_success_rate\": 0,\n",
      "        \"instances\": 0,\n",
      "        \"misclassification_rate\": 0,\n",
      "        \"misclassifications\": 0\n",
      "    },\n",
      "    \"correct\": 1838,\n",
      "    \"test_loss\": 0.3236567614674568\n",
      "} [MainProcess : asyncio_11 (INFO)]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-08-16 23:20:55,060 - <ipython-input-109-7a5cd7c1f6e3>::process(l:36) : Test Outut after Epoch 3 of 10 for Client bladecluster.iitp.org(4) {\n",
      "    \"accuracy\": 89.55,\n",
      "    \"attack\": {\n",
      "        \"attack_success_count\": 0,\n",
      "        \"attack_success_rate\": 0,\n",
      "        \"instances\": 0,\n",
      "        \"misclassification_rate\": 0,\n",
      "        \"misclassifications\": 0\n",
      "    },\n",
      "    \"correct\": 1791,\n",
      "    \"test_loss\": 0.3590500154495239\n",
      "} [MainProcess : asyncio_10 (INFO)]\n",
      "2021-08-16 23:20:55,141 - <ipython-input-109-7a5cd7c1f6e3>::process(l:36) : Test Outut after Epoch 3 of 10 for Client bladecluster.iitp.org(2) {\n",
      "    \"accuracy\": 91.14999999999999,\n",
      "    \"attack\": {\n",
      "        \"attack_success_count\": 0,\n",
      "        \"attack_success_rate\": 0,\n",
      "        \"instances\": 0,\n",
      "        \"misclassification_rate\": 0,\n",
      "        \"misclassifications\": 0\n",
      "    },\n",
      "    \"correct\": 1823,\n",
      "    \"test_loss\": 0.3277583651542664\n",
      "} [MainProcess : asyncio_9 (INFO)]\n",
      "2021-08-16 23:20:55,228 - <ipython-input-109-7a5cd7c1f6e3>::process(l:36) : Test Outut after Epoch 3 of 10 for Client bladecluster.iitp.org(3) {\n",
      "    \"accuracy\": 91.2,\n",
      "    \"attack\": {\n",
      "        \"attack_success_count\": 0,\n",
      "        \"attack_success_rate\": 0,\n",
      "        \"instances\": 0,\n",
      "        \"misclassification_rate\": 0,\n",
      "        \"misclassifications\": 0\n",
      "    },\n",
      "    \"correct\": 1824,\n",
      "    \"test_loss\": 0.326990189909935\n",
      "} [MainProcess : asyncio_6 (INFO)]\n",
      "2021-08-16 23:20:55,255 - <ipython-input-109-7a5cd7c1f6e3>::process(l:36) : Test Outut after Epoch 3 of 10 for Client bladecluster.iitp.org(6) {\n",
      "    \"accuracy\": 90.60000000000001,\n",
      "    \"attack\": {\n",
      "        \"attack_success_count\": 0,\n",
      "        \"attack_success_rate\": 0,\n",
      "        \"instances\": 0,\n",
      "        \"misclassification_rate\": 0,\n",
      "        \"misclassifications\": 0\n",
      "    },\n",
      "    \"correct\": 1812,\n",
      "    \"test_loss\": 0.33620544826984405\n",
      "} [MainProcess : asyncio_8 (INFO)]\n",
      "2021-08-16 23:20:55,649 - <ipython-input-109-7a5cd7c1f6e3>::process(l:36) : Test Outut after Epoch 3 of 10 for Client bladecluster.iitp.org(5) {\n",
      "    \"accuracy\": 90.10000000000001,\n",
      "    \"attack\": {\n",
      "        \"attack_success_count\": 0,\n",
      "        \"attack_success_rate\": 0,\n",
      "        \"instances\": 0,\n",
      "        \"misclassification_rate\": 0,\n",
      "        \"misclassifications\": 0\n",
      "    },\n",
      "    \"correct\": 1802,\n",
      "    \"test_loss\": 0.33495071029663087\n",
      "} [MainProcess : asyncio_7 (INFO)]\n",
      " 30%|███       | 3/10 [01:35<03:47, 32.56s/it]2021-08-16 23:20:55,666 - <ipython-input-110-c8ea6acd90e4>::<module>(l:8) : Federated Training Epoch 4 of 10 [MainProcess : MainThread (INFO)]\n",
      "2021-08-16 23:20:55,684 - <ipython-input-109-7a5cd7c1f6e3>::process(l:12) : Processing Client bladecluster.iitp.org(1) [MainProcess : asyncio_2 (INFO)]\n",
      "2021-08-16 23:20:55,694 - <ipython-input-109-7a5cd7c1f6e3>::process(l:12) : Processing Client bladecluster.iitp.org(3) [MainProcess : asyncio_1 (INFO)]\n",
      "2021-08-16 23:20:55,713 - <ipython-input-109-7a5cd7c1f6e3>::process(l:12) : Processing Client bladecluster.iitp.org(5) [MainProcess : asyncio_0 (INFO)]\n",
      "2021-08-16 23:20:55,721 - <ipython-input-109-7a5cd7c1f6e3>::process(l:12) : Processing Client bladecluster.iitp.org(6) [MainProcess : asyncio_4 (INFO)]\n",
      "2021-08-16 23:20:55,721 - <ipython-input-109-7a5cd7c1f6e3>::process(l:12) : Processing Client bladecluster.iitp.org(2) [MainProcess : asyncio_3 (INFO)]\n",
      "2021-08-16 23:20:55,733 - <ipython-input-109-7a5cd7c1f6e3>::process(l:12) : Processing Client bladecluster.iitp.org(4) [MainProcess : asyncio_5 (INFO)]\n",
      "2021-08-16 23:21:05,756 - <ipython-input-109-7a5cd7c1f6e3>::process(l:16) : Client bladecluster.iitp.org(1) received 6 model updates from ['bladecluster.iitp.org(1)', 'bladecluster.iitp.org(4)', 'bladecluster.iitp.org(2)', 'bladecluster.iitp.org(3)', 'bladecluster.iitp.org(6)', 'bladecluster.iitp.org(5)'] [MainProcess : asyncio_2 (INFO)]\n",
      "2021-08-16 23:21:05,769 - <ipython-input-109-7a5cd7c1f6e3>::process(l:16) : Client bladecluster.iitp.org(3) received 6 model updates from ['bladecluster.iitp.org(1)', 'bladecluster.iitp.org(4)', 'bladecluster.iitp.org(2)', 'bladecluster.iitp.org(3)', 'bladecluster.iitp.org(6)', 'bladecluster.iitp.org(5)'] [MainProcess : asyncio_1 (INFO)]\n",
      "2021-08-16 23:21:05,881 - <ipython-input-109-7a5cd7c1f6e3>::process(l:16) : Client bladecluster.iitp.org(5) received 6 model updates from ['bladecluster.iitp.org(1)', 'bladecluster.iitp.org(4)', 'bladecluster.iitp.org(2)', 'bladecluster.iitp.org(3)', 'bladecluster.iitp.org(6)', 'bladecluster.iitp.org(5)'] [MainProcess : asyncio_0 (INFO)]\n",
      "2021-08-16 23:21:05,881 - <ipython-input-109-7a5cd7c1f6e3>::process(l:16) : Client bladecluster.iitp.org(2) received 6 model updates from ['bladecluster.iitp.org(1)', 'bladecluster.iitp.org(4)', 'bladecluster.iitp.org(2)', 'bladecluster.iitp.org(3)', 'bladecluster.iitp.org(6)', 'bladecluster.iitp.org(5)'] [MainProcess : asyncio_3 (INFO)]\n",
      "2021-08-16 23:21:05,894 - <ipython-input-109-7a5cd7c1f6e3>::process(l:16) : Client bladecluster.iitp.org(4) received 6 model updates from ['bladecluster.iitp.org(1)', 'bladecluster.iitp.org(4)', 'bladecluster.iitp.org(2)', 'bladecluster.iitp.org(3)', 'bladecluster.iitp.org(6)', 'bladecluster.iitp.org(5)'] [MainProcess : asyncio_5 (INFO)]\n",
      "2021-08-16 23:21:05,894 - <ipython-input-109-7a5cd7c1f6e3>::process(l:16) : Client bladecluster.iitp.org(6) received 6 model updates from ['bladecluster.iitp.org(1)', 'bladecluster.iitp.org(4)', 'bladecluster.iitp.org(2)', 'bladecluster.iitp.org(3)', 'bladecluster.iitp.org(6)', 'bladecluster.iitp.org(5)'] [MainProcess : asyncio_4 (INFO)]\n",
      "2021-08-16 23:21:22,639 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:56) : Producing user records to topic pyflx. ^C to exit. [MainProcess : asyncio_4 (INFO)]\n",
      "2021-08-16 23:21:22,827 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:66) : Flushing records... [MainProcess : asyncio_4 (INFO)]\n",
      "2021-08-16 23:21:23,320 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:56) : Producing user records to topic pyflx. ^C to exit. [MainProcess : asyncio_2 (INFO)]\n",
      "2021-08-16 23:21:23,529 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:66) : Flushing records... [MainProcess : asyncio_2 (INFO)]\n",
      "2021-08-16 23:21:23,734 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::delivery_report(l:50) : User record b'bladecluster.iitp.org(6)' successfully produced to pyflx [0] at offset 6 [MainProcess : asyncio_4 (INFO)]\n",
      "2021-08-16 23:21:24,470 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::delivery_report(l:50) : User record b'bladecluster.iitp.org(1)' successfully produced to pyflx [0] at offset 7 [MainProcess : asyncio_2 (INFO)]\n",
      "2021-08-16 23:21:24,470 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:56) : Producing user records to topic pyflx. ^C to exit. [MainProcess : asyncio_0 (INFO)]\n",
      "2021-08-16 23:21:24,679 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:66) : Flushing records... [MainProcess : asyncio_0 (INFO)]\n",
      "2021-08-16 23:21:25,264 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:56) : Producing user records to topic pyflx. ^C to exit. [MainProcess : asyncio_3 (INFO)]\n",
      "2021-08-16 23:21:25,748 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::delivery_report(l:50) : User record b'bladecluster.iitp.org(5)' successfully produced to pyflx [0] at offset 8 [MainProcess : asyncio_0 (INFO)]\n",
      "2021-08-16 23:21:25,761 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:56) : Producing user records to topic pyflx. ^C to exit. [MainProcess : asyncio_5 (INFO)]\n",
      "2021-08-16 23:21:25,883 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:66) : Flushing records... [MainProcess : asyncio_3 (INFO)]\n",
      "2021-08-16 23:21:26,014 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:66) : Flushing records... [MainProcess : asyncio_5 (INFO)]\n",
      "2021-08-16 23:21:26,624 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:56) : Producing user records to topic pyflx. ^C to exit. [MainProcess : asyncio_1 (INFO)]\n",
      "2021-08-16 23:21:26,629 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::delivery_report(l:50) : User record b'bladecluster.iitp.org(2)' successfully produced to pyflx [0] at offset 9 [MainProcess : asyncio_3 (INFO)]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-08-16 23:21:26,791 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::delivery_report(l:50) : User record b'bladecluster.iitp.org(4)' successfully produced to pyflx [0] at offset 10 [MainProcess : asyncio_5 (INFO)]\n",
      "2021-08-16 23:21:26,863 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:66) : Flushing records... [MainProcess : asyncio_1 (INFO)]\n",
      "2021-08-16 23:21:27,691 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::delivery_report(l:50) : User record b'bladecluster.iitp.org(3)' successfully produced to pyflx [0] at offset 11 [MainProcess : asyncio_1 (INFO)]\n",
      "2021-08-16 23:21:28,752 - <ipython-input-109-7a5cd7c1f6e3>::process(l:36) : Test Outut after Epoch 4 of 10 for Client bladecluster.iitp.org(6) {\n",
      "    \"accuracy\": 92.95,\n",
      "    \"attack\": {\n",
      "        \"attack_success_count\": 0,\n",
      "        \"attack_success_rate\": 0,\n",
      "        \"instances\": 0,\n",
      "        \"misclassification_rate\": 0,\n",
      "        \"misclassifications\": 0\n",
      "    },\n",
      "    \"correct\": 1859,\n",
      "    \"test_loss\": 0.2262114907503128\n",
      "} [MainProcess : asyncio_4 (INFO)]\n",
      "2021-08-16 23:21:29,443 - <ipython-input-109-7a5cd7c1f6e3>::process(l:36) : Test Outut after Epoch 4 of 10 for Client bladecluster.iitp.org(1) {\n",
      "    \"accuracy\": 94.25,\n",
      "    \"attack\": {\n",
      "        \"attack_success_count\": 0,\n",
      "        \"attack_success_rate\": 0,\n",
      "        \"instances\": 0,\n",
      "        \"misclassification_rate\": 0,\n",
      "        \"misclassifications\": 0\n",
      "    },\n",
      "    \"correct\": 1885,\n",
      "    \"test_loss\": 0.21970662641525268\n",
      "} [MainProcess : asyncio_2 (INFO)]\n",
      "2021-08-16 23:21:29,689 - <ipython-input-109-7a5cd7c1f6e3>::process(l:36) : Test Outut after Epoch 4 of 10 for Client bladecluster.iitp.org(5) {\n",
      "    \"accuracy\": 93.95,\n",
      "    \"attack\": {\n",
      "        \"attack_success_count\": 0,\n",
      "        \"attack_success_rate\": 0,\n",
      "        \"instances\": 0,\n",
      "        \"misclassification_rate\": 0,\n",
      "        \"misclassifications\": 0\n",
      "    },\n",
      "    \"correct\": 1879,\n",
      "    \"test_loss\": 0.21235963299870492\n",
      "} [MainProcess : asyncio_0 (INFO)]\n",
      "2021-08-16 23:21:29,830 - <ipython-input-109-7a5cd7c1f6e3>::process(l:36) : Test Outut after Epoch 4 of 10 for Client bladecluster.iitp.org(4) {\n",
      "    \"accuracy\": 93.15,\n",
      "    \"attack\": {\n",
      "        \"attack_success_count\": 0,\n",
      "        \"attack_success_rate\": 0,\n",
      "        \"instances\": 0,\n",
      "        \"misclassification_rate\": 0,\n",
      "        \"misclassifications\": 0\n",
      "    },\n",
      "    \"correct\": 1863,\n",
      "    \"test_loss\": 0.23099512749910356\n",
      "} [MainProcess : asyncio_5 (INFO)]\n",
      "2021-08-16 23:21:29,890 - <ipython-input-109-7a5cd7c1f6e3>::process(l:36) : Test Outut after Epoch 4 of 10 for Client bladecluster.iitp.org(2) {\n",
      "    \"accuracy\": 93.0,\n",
      "    \"attack\": {\n",
      "        \"attack_success_count\": 0,\n",
      "        \"attack_success_rate\": 0,\n",
      "        \"instances\": 0,\n",
      "        \"misclassification_rate\": 0,\n",
      "        \"misclassifications\": 0\n",
      "    },\n",
      "    \"correct\": 1860,\n",
      "    \"test_loss\": 0.2195892654657364\n",
      "} [MainProcess : asyncio_3 (INFO)]\n",
      "2021-08-16 23:21:30,172 - <ipython-input-109-7a5cd7c1f6e3>::process(l:36) : Test Outut after Epoch 4 of 10 for Client bladecluster.iitp.org(3) {\n",
      "    \"accuracy\": 93.85,\n",
      "    \"attack\": {\n",
      "        \"attack_success_count\": 0,\n",
      "        \"attack_success_rate\": 0,\n",
      "        \"instances\": 0,\n",
      "        \"misclassification_rate\": 0,\n",
      "        \"misclassifications\": 0\n",
      "    },\n",
      "    \"correct\": 1877,\n",
      "    \"test_loss\": 0.21329003047943115\n",
      "} [MainProcess : asyncio_1 (INFO)]\n",
      " 40%|████      | 4/10 [02:09<03:20, 33.34s/it]2021-08-16 23:21:30,192 - <ipython-input-110-c8ea6acd90e4>::<module>(l:8) : Federated Training Epoch 5 of 10 [MainProcess : MainThread (INFO)]\n",
      "2021-08-16 23:21:30,241 - <ipython-input-109-7a5cd7c1f6e3>::process(l:12) : Processing Client bladecluster.iitp.org(1) [MainProcess : asyncio_10 (INFO)]\n",
      "2021-08-16 23:21:30,252 - <ipython-input-109-7a5cd7c1f6e3>::process(l:12) : Processing Client bladecluster.iitp.org(5) [MainProcess : asyncio_7 (INFO)]\n",
      "2021-08-16 23:21:30,258 - <ipython-input-109-7a5cd7c1f6e3>::process(l:12) : Processing Client bladecluster.iitp.org(2) [MainProcess : asyncio_9 (INFO)]\n",
      "2021-08-16 23:21:30,260 - <ipython-input-109-7a5cd7c1f6e3>::process(l:12) : Processing Client bladecluster.iitp.org(4) [MainProcess : asyncio_8 (INFO)]\n",
      "2021-08-16 23:21:30,266 - <ipython-input-109-7a5cd7c1f6e3>::process(l:12) : Processing Client bladecluster.iitp.org(3) [MainProcess : asyncio_6 (INFO)]\n",
      "2021-08-16 23:21:30,269 - <ipython-input-109-7a5cd7c1f6e3>::process(l:12) : Processing Client bladecluster.iitp.org(6) [MainProcess : asyncio_11 (INFO)]\n",
      "2021-08-16 23:21:40,324 - <ipython-input-109-7a5cd7c1f6e3>::process(l:16) : Client bladecluster.iitp.org(5) received 6 model updates from ['bladecluster.iitp.org(6)', 'bladecluster.iitp.org(1)', 'bladecluster.iitp.org(5)', 'bladecluster.iitp.org(2)', 'bladecluster.iitp.org(4)', 'bladecluster.iitp.org(3)'] [MainProcess : asyncio_7 (INFO)]\n",
      "2021-08-16 23:21:40,324 - <ipython-input-109-7a5cd7c1f6e3>::process(l:16) : Client bladecluster.iitp.org(2) received 6 model updates from ['bladecluster.iitp.org(6)', 'bladecluster.iitp.org(1)', 'bladecluster.iitp.org(5)', 'bladecluster.iitp.org(2)', 'bladecluster.iitp.org(4)', 'bladecluster.iitp.org(3)'] [MainProcess : asyncio_9 (INFO)]\n",
      "2021-08-16 23:21:40,335 - <ipython-input-109-7a5cd7c1f6e3>::process(l:16) : Client bladecluster.iitp.org(1) received 6 model updates from ['bladecluster.iitp.org(6)', 'bladecluster.iitp.org(1)', 'bladecluster.iitp.org(5)', 'bladecluster.iitp.org(2)', 'bladecluster.iitp.org(4)', 'bladecluster.iitp.org(3)'] [MainProcess : asyncio_10 (INFO)]\n",
      "2021-08-16 23:21:40,347 - <ipython-input-109-7a5cd7c1f6e3>::process(l:16) : Client bladecluster.iitp.org(4) received 6 model updates from ['bladecluster.iitp.org(6)', 'bladecluster.iitp.org(1)', 'bladecluster.iitp.org(5)', 'bladecluster.iitp.org(2)', 'bladecluster.iitp.org(4)', 'bladecluster.iitp.org(3)'] [MainProcess : asyncio_8 (INFO)]\n",
      "2021-08-16 23:21:40,347 - <ipython-input-109-7a5cd7c1f6e3>::process(l:16) : Client bladecluster.iitp.org(6) received 6 model updates from ['bladecluster.iitp.org(6)', 'bladecluster.iitp.org(1)', 'bladecluster.iitp.org(5)', 'bladecluster.iitp.org(2)', 'bladecluster.iitp.org(4)', 'bladecluster.iitp.org(3)'] [MainProcess : asyncio_11 (INFO)]\n",
      "2021-08-16 23:21:40,407 - <ipython-input-109-7a5cd7c1f6e3>::process(l:16) : Client bladecluster.iitp.org(3) received 6 model updates from ['bladecluster.iitp.org(6)', 'bladecluster.iitp.org(1)', 'bladecluster.iitp.org(5)', 'bladecluster.iitp.org(2)', 'bladecluster.iitp.org(4)', 'bladecluster.iitp.org(3)'] [MainProcess : asyncio_6 (INFO)]\n",
      "2021-08-16 23:21:57,591 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:56) : Producing user records to topic pyflx. ^C to exit. [MainProcess : asyncio_11 (INFO)]\n",
      "2021-08-16 23:21:57,960 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:56) : Producing user records to topic pyflx. ^C to exit. [MainProcess : asyncio_7 (INFO)]\n",
      "2021-08-16 23:21:58,041 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:66) : Flushing records... [MainProcess : asyncio_11 (INFO)]\n",
      "2021-08-16 23:21:58,172 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:66) : Flushing records... [MainProcess : asyncio_7 (INFO)]\n",
      "2021-08-16 23:21:58,429 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:56) : Producing user records to topic pyflx. ^C to exit. [MainProcess : asyncio_6 (INFO)]\n",
      "2021-08-16 23:21:58,793 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::delivery_report(l:50) : User record b'bladecluster.iitp.org(6)' successfully produced to pyflx [0] at offset 12 [MainProcess : asyncio_11 (INFO)]\n",
      "2021-08-16 23:21:58,799 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:56) : Producing user records to topic pyflx. ^C to exit. [MainProcess : asyncio_9 (INFO)]\n",
      "2021-08-16 23:21:58,970 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:66) : Flushing records... [MainProcess : asyncio_6 (INFO)]\n",
      "2021-08-16 23:21:59,298 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::delivery_report(l:50) : User record b'bladecluster.iitp.org(5)' successfully produced to pyflx [0] at offset 13 [MainProcess : asyncio_7 (INFO)]\n",
      "2021-08-16 23:21:59,333 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:56) : Producing user records to topic pyflx. ^C to exit. [MainProcess : asyncio_10 (INFO)]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-08-16 23:21:59,448 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:66) : Flushing records... [MainProcess : asyncio_9 (INFO)]\n",
      "2021-08-16 23:21:59,498 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::delivery_report(l:50) : User record b'bladecluster.iitp.org(3)' successfully produced to pyflx [0] at offset 14 [MainProcess : asyncio_6 (INFO)]\n",
      "2021-08-16 23:21:59,600 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:66) : Flushing records... [MainProcess : asyncio_10 (INFO)]\n",
      "2021-08-16 23:22:00,020 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::delivery_report(l:50) : User record b'bladecluster.iitp.org(2)' successfully produced to pyflx [0] at offset 15 [MainProcess : asyncio_9 (INFO)]\n",
      "2021-08-16 23:22:00,022 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:56) : Producing user records to topic pyflx. ^C to exit. [MainProcess : asyncio_8 (INFO)]\n",
      "2021-08-16 23:22:00,210 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::produce(l:66) : Flushing records... [MainProcess : asyncio_8 (INFO)]\n",
      "2021-08-16 23:22:00,377 - /home/harsh_1921cs01/hub/F3IA/fl/libs/protobuf_producer.py::delivery_report(l:50) : User record b'bladecluster.iitp.org(1)' successfully produced to pyflx [0] at offset 16 [MainProcess : asyncio_10 (INFO)]\n",
      " 40%|████      | 4/10 [02:40<04:00, 40.04s/it]\n"
     ]
    },
    {
     "ename": "CancelledError",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mCancelledError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-110-c8ea6acd90e4>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     13\u001b[0m                      \u001b[0mclients_info\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mclient\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'test_loader'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     14\u001b[0m                      fedargs, device) for client in clients]\n\u001b[0;32m---> 15\u001b[0;31m     \u001b[0;32mawait\u001b[0m \u001b[0masyncio\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtasks\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     16\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mstart_time\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/syft/lib/python3.9/asyncio/tasks.py\u001b[0m in \u001b[0;36mwait\u001b[0;34m(fs, loop, timeout, return_when)\u001b[0m\n\u001b[1;32m    411\u001b[0m     \u001b[0mfs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0mensure_future\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mloop\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mloop\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mf\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mfs\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    412\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 413\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0;32mawait\u001b[0m \u001b[0m_wait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreturn_when\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mloop\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    414\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    415\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/syft/lib/python3.9/asyncio/tasks.py\u001b[0m in \u001b[0;36m_wait\u001b[0;34m(fs, timeout, return_when, loop)\u001b[0m\n\u001b[1;32m    527\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    528\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 529\u001b[0;31m         \u001b[0;32mawait\u001b[0m \u001b[0mwaiter\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    530\u001b[0m     \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    531\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mtimeout_handle\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/syft/lib/python3.9/asyncio/futures.py\u001b[0m in \u001b[0;36m__await__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    282\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdone\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    283\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_asyncio_future_blocking\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 284\u001b[0;31m             \u001b[0;32myield\u001b[0m \u001b[0mself\u001b[0m  \u001b[0;31m# This tells Task to wait for completion.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    285\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdone\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    286\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mRuntimeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"await wasn't used with future\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/syft/lib/python3.9/asyncio/tasks.py\u001b[0m in \u001b[0;36m__wakeup\u001b[0;34m(self, future)\u001b[0m\n\u001b[1;32m    326\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__wakeup\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfuture\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    327\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 328\u001b[0;31m             \u001b[0mfuture\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    329\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mBaseException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mexc\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    330\u001b[0m             \u001b[0;31m# This may also be a cancellation.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/syft/lib/python3.9/asyncio/futures.py\u001b[0m in \u001b[0;36mresult\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    194\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_state\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0m_CANCELLED\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    195\u001b[0m             \u001b[0mexc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_cancelled_error\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 196\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mexc\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    197\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_state\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0m_FINISHED\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    198\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mexceptions\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mInvalidStateError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Result is not ready.'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mCancelledError\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import time\n",
    "start_time = time.time()\n",
    "\n",
    "# Federated Training\n",
    "for _epoch in tqdm(range(fedargs.epochs)):\n",
    "\n",
    "    epoch = _epoch + 1\n",
    "    log.info(\"Federated Training Epoch {} of {}\".format(epoch, fedargs.epochs))\n",
    "\n",
    "    # Clients  \n",
    "    tasks = [process(client, epoch, ctp, client_models[client],\n",
    "                     clients_info[client]['train_loader'],\n",
    "                     clients_info[client]['test_loader'],\n",
    "                     fedargs, device) for client in clients]\n",
    "    await asyncio.wait(tasks)\n",
    "    \n",
    "print(time.time() - start_time)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfa0923e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60cee49d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
